larq/optimizers_v2.py

```diff
@@ -1,10 +1,121 @@
 
 from larq import utils
 from copy import deepcopy
 
 
+@utils.register_keras_custom_object
+class CaseOptimizer(tf.keras.optimizers.Optimizer):
+    """A case operation for optimizers that each train a subset of a model's variables.
+
+    An optimizer is used to train a variable iff its accompanying predicate evaluates to
+    `True`.
+
+    For each variable, at most one optimizer's predicate may evaluate to `True`. If no
+    optimizer's predicate evaluates to `True` for a variable, it is trained with the
+    `default` optimizer.
+
+    # Arguments
+    pred_opt_pairs: a list of `(tf.keras.optimizers.Optimzer, pred)` pairs, where `pred` 
+        takes one `tf.Variable` as argument and returns `True` if the optimizer should
+        train that variable, e.g. `pred(var) === True`.
+    default: a `tf.keras.optimizers.Optimizer` to be applied to any variable not claimed
+        by any other optimizer.
+    """
+
+    def __init__(self, pred_opt_pairs, default, name="optim_case"):
+        super().__init__(name=name)
+
+        for i, (pred, opt) in enumerate(pred_opt_pairs):
+            if not callable(pred):
+                raise TypeError(
+                    f"Expected callable predicate at `pred_opt_pairs[{i}][0]` but got `{type(pred)}`."
+                )
+            if not isinstance(opt, tf.keras.optimizers.Optimizer):
+                raise TypeError(
+                    f"Expected `tf.keras.optimizers.Optimizer` at `pred_opt_pairs[{i}][1]` but got `{type(opt)}`."
+                )
+        self.pred_opt_pairs = pred_opt_pairs
+
+        if default is not None and not isinstance(
+            default, tf.keras.optimizers.Optimizer
+        ):
+            raise TypeError(
+                f"Expected `tf.keras.optimizers.Optimizer` for `default` but got `{type(fallback_optimizer)}`."
+            )
+        self.default = default
+
+    def __getattr__(self, name):
+        if name == "lr":
+            raise ValueError("No single learning rate available for CaseOptimizer.")
+        return super().__getattr__(name)
+
+    def apply_gradients(self, grads_and_vars, name=None):
+        opt_grads_and_vars = [[] for _ in range(len(self.pred_opt_pairs))]
+        default_grads_and_vars = []
+
+        for grad, var in grads_and_vars:
+            num_opts = 0
+            for i, (pred, opt) in enumerate(self.pred_opt_pairs):
+                if pred(var):
+                    opt_grads_and_vars[i].append((grad, var))
+                    num_opts += 1
+
+            if num_opts == 0:
+                default_grads_and_vars.append((grad, var))
+            if num_opts > 1:
+                raise ValueError(f"Variable `{var}` claimed by multiple optimizers.")
+
+        train_ops = []
+        for (_, opt), grads_and_vars in zip(self.pred_opt_pairs, opt_grads_and_vars):
+            train_ops.append(opt.apply_gradients(grads_and_vars, name=name))
+
+        if len(default_grads_and_vars) > 0 and self.default is None:
```

Comment:
If we throw at line 68, we could remove this logic.

Reply:
See [my comment above](https://github.com/larq/larq/pull/304#discussion_r343012326), we don't necessarily want to throw.